dogu/versionRegistry.go

```diff
@@ -0,0 +1,423 @@
+	logger := log.FromContext(ctx).WithName("VersionRegistry.waitForWatchEvents")
+	for {
+		select {
+		case <-ctx.Done():
+			watchInterface.Stop()
+			logger.Info("context canceled. Stop watch channel.")
+			return
+		case event, open := <-watchInterface.ResultChan():
+			if !open {
+				logger.Info("watch channel canceled. Stop watch.")
+				return
+			}
+			switch event.Type {
+			case watch.Added:
+				err := handleAddWatchEvent(event, persistenceContext, currentVersionsWatchResult)
+				if err != nil {
+					throwAndLogWatchError(ctx, fmt.Errorf("failed to handle add watch event: %w", err), currentVersionsWatchResult)
+				}
+			case watch.Modified:
+				err := handleModifiedWatchEvent(ctx, event, persistenceContext, currentVersionsWatchResult)
+				if err != nil {
+					throwAndLogWatchError(ctx, fmt.Errorf("failed to handle modified watch event: %w", err), currentVersionsWatchResult)
+				}
+			case watch.Deleted:
+				err := handleDeleteWatchEvent(event, persistenceContext, currentVersionsWatchResult)
+				if err != nil {
+					throwAndLogWatchError(ctx, fmt.Errorf("failed to handle delete watch event: %w", err), currentVersionsWatchResult)
+				}
+			case watch.Error:
+				status, ok := event.Object.(*metav1.Status)
+				if !ok {
+					throwAndLogWatchError(ctx, fmt.Errorf("failed to cast event object to %T", metav1.Status{}), currentVersionsWatchResult)
+					break
+				}
+				throwAndLogWatchError(ctx, fmt.Errorf("watch event type is error: %q", status.String()), currentVersionsWatchResult)
+				return
+			}
+		}
+	}
+}
+
+func handleDeleteWatchEvent(event watch.Event, persistenceContext map[SimpleDoguName]core.Version, currentVersionsWatchResult chan CurrentVersionsWatchResult) error {
+	specConfigMap, err := getSpecConfigMapFromEvent(event)
+	if err != nil {
+		return err
+	}
+
+	if !hasDoguSpecConfigMapCurrentKey(specConfigMap) {
+		// disabled dogus deleted. Do nothing
+		return nil
+	}
+
+	eventDoguVersion, err := getCurrentDoguVersionFromDoguSpecConfigMap(*specConfigMap)
+	if err != nil {
+		return err
+	}
+
+	oldPersistenceContext := copyPersistenceContext(persistenceContext)
+	delete(persistenceContext, eventDoguVersion.Name)
+
+	fireWatchResult(currentVersionsWatchResult, oldPersistenceContext, persistenceContext, []DoguVersion{eventDoguVersion})
+	return nil
+}
+
+func handleModifiedWatchEvent(ctx context.Context, event watch.Event, persistenceContext map[SimpleDoguName]core.Version, currentVersionsWatchResult chan CurrentVersionsWatchResult) error {
+	logger := log.FromContext(ctx).WithName("VersionRegistry.handleModifiedWatchEvent")
+	specConfigMap, err := getSpecConfigMapFromEvent(event)
+	if err != nil {
+		return err
+	}
+
+	oldPersistenceContext := copyPersistenceContext(persistenceContext)
+
+	// Skip process. Configmap was possible created empty and will get modified event on Enable.
```

Comment:
This is the case if the dogu-operator install a new dogu. First he will create the cm but possible without any content. This event should be skipped. After that he will update the configmap with the new dogu entry. This will be handled by a modified event.